cluster/src/main/java/org/apache/iotdb/cluster/client/sync/SyncDataClient.java

```diff
@@ -19,64 +19,142 @@
     super(prot);
   }
 
-  SyncDataClient(TProtocolFactory protocolFactory, Node target, SyncClientPool pool)
+  public SyncDataClient(TProtocolFactory protocolFactory, Node node, ClientCategory category)
       throws TTransportException {
+
+    // the difference of the two clients lies in the port
     super(
-        protocolFactory,
-        target.getInternalIp(),
-        target.getDataPort(),
-        ClusterConstant.getConnectionTimeoutInMS(),
-        target,
-        pool);
+        protocolFactory.getProtocol(
+            RpcTransportFactory.INSTANCE.getTransport(
+                new TSocket(
+                    TConfigurationConst.defaultTConfiguration,
+                    node.getInternalIp(),
+                    ClientUtils.getPort(node, category),
+                    ClusterConstant.getConnectionTimeoutInMS()))));
+    this.node = node;
+    this.category = category;
+    getInputProtocol().getTransport().open();
+  }
+
+  public void setClientPool(IClientPool clientPool) {
+    this.clientPool = clientPool;
+  }
+
+  public void returnSelf() {
+    if (clientPool != null) clientPool.returnSyncClient(this, node, category);
+  }
+
+  public void setTimeout(int timeout) {
+    // the same transport is used in both input and output
+    ((TimeoutChangeableTransport) (getInputProtocol().getTransport())).setTimeout(timeout);
+  }
+
+  public void close() {
+    getInputProtocol().getTransport().close();
+  }
+
+  @TestOnly
+  public int getTimeout() throws SocketException {
+    return ((TimeoutChangeableTransport) getInputProtocol().getTransport()).getTimeOut();
   }
 
   @Override
   public String toString() {
-    return String.format(
-        "SyncDataClient (ip = %s, port = %d, id = %d)",
-        target.getInternalIp(), target.getDataPort(), target.getNodeIdentifier());
+    return "Sync"
+        + category.getName()
+        + "{"
+        + "node="
+        + node
+        + ","
+        + "port="
+        + ClientUtils.getPort(node, category)
+        + '}';
+  }
+
+  public Node getNode() {
+    return node;
   }
 
-  public static class Factory implements SyncClientFactory {
+  public static class SyncDataClientFactory
+      implements KeyedPooledObjectFactory<Node, SyncDataClient> {
 
     private TProtocolFactory protocolFactory;
+    private ClientCategory category;
 
-    public Factory(TProtocolFactory protocolFactory) {
+    public SyncDataClientFactory(TProtocolFactory protocolFactory, ClientCategory category) {
       this.protocolFactory = protocolFactory;
+      this.category = category;
+    }
+
+    //  public String nodeInfo(Node node) {
+    //    return String.format(
+    //        "MetaNode (listenIp = %s, HB port = %d, id = %d)",
+    //        node.getInternalIp(),
+    //        node.getMetaPort() + ClusterUtils.DATA_HEARTBEAT_PORT_OFFSET,
+    //        node.getNodeIdentifier());
+    //  }
+
+    @Override
+    public void activateObject(Node node, PooledObject<SyncDataClient> pooledObject)
```

Comment:
Do we have to override these functions with empty implementation?

Reply:
`activateObject` is called before the object is borrow by come application from pool to do some initialization. Leave it empty is ok for thrift client as we have open transport when do construct and make it validate when the object is returned. I will do some re-initialization here.

Reply:
I see~